src/expr/parser.rs

```diff
@@ -1,58 +1,76 @@
-// Alex Ozdemir <aozdemir@hmc.edu> // <- Your name should replace this line!
+// Dan Obermiller <dobermiller16@cmc.edu>
 // Starter code for HMC's MemorySafe, week 2
 //
 // The parser for an `Expr` (currently just produces the value fo the `Expr`)
 
 use nom::digit;
 
+use expr::{Expr, BinOp};
 use std::str;
+use std::mem;
 use std::str::FromStr;
 
-named!(parens<i64>, delimited!(
+named!(parens<Expr>, delimited!(
     char!('('),
     expr,
     char!(')')
   )
 );
 
-named!(i64_literal<i64>,
-  map_res!(
+named!(i64_literal<Expr>,
+  map!(
     map_res!(
-      digit,
-      str::from_utf8
+      map_res!(
+        digit,
+        str::from_utf8
+      ),
+      FromStr::from_str
     ),
-    FromStr::from_str
+    Expr::Literal
   )
 );
 
-named!(factor<i64>,
+named!(factor<Expr>,
   alt!(
     i64_literal
   | parens
   )
 );
 
+fn lambda_y(acc: &mut Expr, op: BinOp, e: Expr) {
+    let tmp = mem::replace(acc, Expr::Literal(0));
+    mem::replace(acc, Expr::BinOp(Box::new(tmp), op, Box::new(e)));
+}
```

Comment:
Nice job abstracting this out!

It could probably use a better name and be a method on `Expr`.